pkg/apis/pipeline/v1alpha1/cloud_event_resource.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2019 The Tekton Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// CloudEventResource is an event sink to which events are delivered when a TaskRun has finished
+type CloudEventResource struct {
+	// Name is the name used to reference to the PipelineResource
+	Name      string               `json:"name"`
+	// Type must be `PipelineResourceTypeCloudEvent`
+	Type      PipelineResourceType `json:"type"`
+	// TargetURI is the URI of the sink which the cloud event is develired to
+	TargetURI string               `json:"targetURI"`
+}
+
+// NewCloudEventResource creates a new CloudEvent resource to pass to a Task
+func NewCloudEventResource(r *PipelineResource) (*CloudEventResource, error) {
+	if r.Spec.Type != PipelineResourceTypeCloudEvent {
+		return nil, fmt.Errorf("CloudEventResource: Cannot create a Cloud Event resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	var targetURI string
+	var targetURISpecified bool
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "TargetURI"):
+			targetURI = param.Value
+			if param.Value != "" {
+				targetURISpecified = true
+			}
+		}
+	}
+
+	if !targetURISpecified {
+		return nil, fmt.Errorf("CloudEventResource: Need URI to be specified in order to create a CloudEvent resource %s", r.Name)
+	}
+	return &CloudEventResource{
+		Name:      r.Name,
+		Type:      r.Spec.Type,
+		TargetURI: targetURI,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s CloudEventResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s CloudEventResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeCloudEvent
+}
+
+// Replacements is used for template replacement on an CloudEventResource inside of a Taskrun.
+func (s *CloudEventResource) Replacements() map[string]string {
+	return map[string]string{
+		"name":       s.Name,
+		"type":       string(s.Type),
+		"target-uri": s.TargetURI,
+	}
+}
+
+// GetUploadContainerSpec returns nothing as the cloud event is sent by the controller once the POD execution is completed
+func (s *CloudEventResource) GetUploadContainerSpec() ([]corev1.Container, error) {
+	return nil, nil
+}
```

Comment:
Yeah, I wanted to do this using a container scheduled to be executed as the last one in the Pod, but it didn't work out, so we might switch this in future to be something like you proposed.

I like both the ideas you proposed, I think what I need to do is to put together a design for a more general concept of notifications

One thing that I wanted to highlight is that the way this will be used in https://github.com/tektoncd/pipeline/pull/837 is for TaskRun notifications, as opposed to PipelineRun. 
Both TaskRun and PipelineRun end notifications can be useful, but I focused on TaskRun first because it fits well with the use cases I included in the design.

In the demo that I presented at the OSS I have a pipeline that builds a docker image, runs unit testing and runs a code lint, the three in parallel. All three tasks send a notifications once they're done, and the notification triggers different TaskRuns depending on the source (identified via a label on the TaskRun). This makes it possible to publish test results on GitHub using the PullRequestPipelineResource as soon as the corresponding task is complete, with no need to wait for the other Tasks - i.e. I don't necessarily want to wait for the image build to be complete before I can see the unit test results.